src/core/agent/services/multiSigService.ts

```diff
@@ -0,0 +1,710 @@
+    const identifier = identifiers.find((identifier) => {
+      return smids.find((member) => identifier.id === member);
+    });
+
+    if (!identifier) {
+      throw new Error(MultiSigService.CANNOT_JOIN_MULTISIG_ICP);
+    }
+
+    const aid = await this.signifyClient
+      .identifiers()
+      .get(identifier?.signifyName);
+    const signifyName = uuidv4();
+    const res = await this.joinMultisigKeri(exn, aid, signifyName);
+    await this.basicStorage.deleteById(notification.id);
+    const multisigId = res.op.name.split(".")[1];
+    await this.identifierStorage.createIdentifierMetadataRecord({
+      id: multisigId,
+      displayName: meta.displayName,
+      colors: meta.colors,
+      theme: meta.theme,
+      signifyName,
+      signifyOpName: res.op.name, //we save the signifyOpName here to sync the multisig's status later
+      isPending: res.op.done ? false : true, //this will be updated once the operation is done
+      multisigManageAid: identifier.id,
+    });
+    return { identifier: multisigId, signifyName };
+  }
+
+  async markMultisigCompleteIfReady(metadata: IdentifierMetadataRecord) {
+    if (!metadata.signifyOpName || !metadata.isPending) {
+      return {
+        done: true,
+      };
+    }
+    const pendingOperation = await this.signifyClient
+      .operations()
+      .get(metadata.signifyOpName);
+    if (pendingOperation && pendingOperation.done) {
+      await this.identifierStorage.updateIdentifierMetadata(metadata.id, {
+        isPending: false,
+      });
+      return { done: true };
+    }
+    return { done: false };
+  }
+
+  async getUnhandledMultisigIdentifiers(
+    filters: {
+      isDismissed?: boolean;
+    } = {}
+  ): Promise<KeriNotification[]> {
+    const results = await this.basicStorage.findAllByQuery(
+      RecordType.NOTIFICATION_KERI,
+      {
+        route: NotificationRoute.MultiSigIcp,
+        ...filters,
+        $or: [
+          { route: NotificationRoute.MultiSigIcp },
+          {
+            route: NotificationRoute.MultiSigRot,
+          },
+        ],
+      }
+    );
+    return results.map((result) => {
+      return {
+        id: result.id,
+        createdAt: result.createdAt,
+        a: result.content,
+      };
+    });
+  }
+
+  async createDelegatedIdentifier(
+    metadata: Omit<
+      IdentifierMetadataRecordProps,
+      "id" | "createdAt" | "isArchived" | "signifyName"
+    >,
+    delegatorPrefix: string
+  ): Promise<CreateIdentifierResult> {
+    const signifyName = uuidv4();
+    const operation = await this.signifyClient
+      .identifiers()
+      .create(signifyName, { delpre: delegatorPrefix });
+    const identifier = operation.serder.ked.i;
+    await this.identifierStorage.createIdentifierMetadataRecord({
+      id: identifier,
+      ...metadata,
+      signifyName: signifyName,
+      isPending: true,
+    });
+    return { identifier, signifyName };
+  }
+
+  async approveDelegation(
```

Comment:
There's quite a few delegation functions too - maybe this can be its own service?

Reply:
Hmm, don't fully understand your question, but I think it is function of identifiderService. Need remove to it?

Reply:
@bao-sotatek `approveDelegation` is related to delegation, not multi-sig. This is a multi-sig file. Delegation should be in its own file.

The only delegation that should be in multisig is delegated multi-sig.